### PR TITLE
[feat] - Refactor AzureSD

### PR DESCRIPTION
**Pull request by heliapb**
## Description

From issue https://github.com/prometheus-operator/prometheus-operator/issues/7202

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
ENHANCEMENT: Refactor AzureSD
```

### CODE REVIEW

**Review by simonpasquier — DISMISSED** [review]
thanks!

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 1]
```diff
@@ -692,7 +692,7 @@ type AzureSDConfig struct {
 	// The port to scrape metrics from. If using the public IP address, this must
 	// instead be specified in the relabeling rule.
 	// +optional
-	Port *int `json:"port"`
+	Port *int `json:"port,omitempty"`
```
Comment (slashpai) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:695]: I think we can include minimum and maximum value for port number like did in EC2SDConfig
cc: @mviswanathsai

Comment (heliapb) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: thanks, fixed

**Comment by slashpai** [comment]
A few more fields still need minlength validation, `// +kubebuilder:validation:MinLength=1`

```go
        // Optional tenant ID. Only required with the OAuth authentication method.
	// +optional
	TenantID *string `json:"tenantID,omitempty"`
	// Optional client ID. Only required with the OAuth authentication method.
	// +optional
	ClientID *string `json:"clientID,omitempty"`
	// Optional resource group name. Limits discovery to this resource group.
	// +optional
	ResourceGroup *string `json:"resourceGroup,omitempty"`
```

**Comment by mviswanathsai** [comment]
Adding to what Jayapriya said,
```
	// The Azure environment.
        // +kubebuilder:validation:MinLength=1
	// +optional
	Environment *string `json:"environment,omitempty"`
```

**Comment by mviswanathsai** [comment]
```
AuthenticationMethod *AuthenticationMethodType `json...`
```
And 
```
// +kubebuilder:validation:Enum=OAuth;ManagedIdentity;SDK
type AuthenticationMethodType string

const (
AuthMethodTypeOAuth AuthenticationMethodType = "OAuth"
// and so on, for the rest of the types
)
```
This makes the code more reusable, we can use these constants in our tests instead of using strings and risking typos. Similar to what we have with `DNSRecordTypes`.

**Comment by mviswanathsai** [comment]
Just as a heads up, looking at the `UnmarshalYAML()` function of each of the service discoveries gives a good idea of what is expected from each field. In this case, https://github.com/prometheus/prometheus/blob/main/discovery/azure/azure.go#L146

File: test/e2e/scrapeconfig_test.go [hunk 2]
```diff
@@ -2035,3 +2038,107 @@ var IonosSDTestCases = []scrapeCRDTestCase{
...
+				{
+					Environment:          ptr.To("AzurePublicCloud"),
+					AuthenticationMethod: ptr.To("OAuth"),
+					SubscriptionID:       "11111111-1111-1111-1111-111111111111",
+					RefreshInterval:      ptr.To(monitoringv1.Duration("30g")),
+				},
+			},
+		},
+		expectedError: true,
+	},
```
Comment (mviswanathsai) [inline test/e2e/scrapeconfig_test.go:2143]: Could you also add tests for the other optional fields, checking for the correct working of the min length length validation?

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 3]
```diff
@@ -675,24 +675,29 @@ type AzureSDConfig struct {
...
 	// +optional
 	TenantID *string `json:"tenantID,omitempty"`
 	// Optional client ID. Only required with the OAuth authentication method.
+	// +kubebuilder:validation:MinLength=1
 	// +optional
 	ClientID *string `json:"clientID,omitempty"`
 	// Optional client secret. Only required with the OAuth authentication method.
 	// +optional
 	ClientSecret *corev1.SecretKeySelector `json:"clientSecret,omitempty"`
 	// Optional resource group name. Limits discovery to this resource group.
```
Comment (mviswanathsai) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:688]:
```suggestion
	// Optional resource group name. Limits discovery to this resource group.
	// Requires  Prometheus v2.35.0 and above
```
Also, we need to check for the version and reject any object that fails this check at `resource_selector.go`

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 4]
```diff
@@ -675,24 +675,29 @@ type AzureSDConfig struct {
...
 	// RefreshInterval configures the refresh interval at which Prometheus will re-read the instance list.
 	// +optional
 	RefreshInterval *v1.Duration `json:"refreshInterval,omitempty"`
 	// The port to scrape metrics from. If using the public IP address, this must
 	// instead be specified in the relabeling rule.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
 	// +optional
-	Port *int `json:"port"`
+	Port *int32 `json:"port,omitempty"`
```
Comment (mviswanathsai) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:700]: Azure also supports HTTP client configuration. See https://prometheus.io/docs/prometheus/latest/configuration/configuration/#azure_sd_config

You would need to add something similar to:
```Go
	// BasicAuth information to authenticate against the target HTTP endpoint.
	// More info: https://prometheus.io/docs/operating/configuration/#endpoints
	// Cannot be set at the same time as `authorization`, or `oAuth2`.
	// +optional
	BasicAuth *v1.BasicAuth `json:"basicAuth,omitempty"`
	// Authorization header configuration to authenticate against the target HTTP endpoint.
	// Cannot be set at the same time as `oAuth2`, or `basicAuth`.
	// +optional
	Authorization *v1.SafeAuthorization `json:"authorization,omitempty"`
	// Optional OAuth 2.0 configuration to authenticate against the target HTTP endpoint.
	// Cannot be set at the same time as `authorization`, or `basicAuth`.
	// +optional
	OAuth2         *v1.OAuth2 `json:"oauth2,omitempty"`
	v1.ProxyConfig `json:",inline"`
	// TLS configuration applying to the target HTTP endpoint.
	// +optional
	TLSConfig *v1.SafeTLSConfig `json:"tlsConfig,omitempty"`
	// Configure whether HTTP requests follow HTTP 3xx redirects.
	// +optional
	FollowRedirects *bool `json:"followRedirects,omitempty"`
	// Whether to enable HTTP2.
	// +optional
	EnableHTTP2 *bool `json:"enableHTTP2,omitempty"`
```

Comment (mviswanathsai) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:700]: HTTP client config for AzureSD is supported only for Prometheus >= v2.31.0. So I would suggest we also add the rejection logic at `resource_selector.go` for this.

File: test/e2e/scrapeconfig_test.go (hunk 2 above)
Comment (mviswanathsai) [reply test/e2e/scrapeconfig_test.go:2143]: Also, once the version checks are added I would also like to see some versioned testcases to see that the object is being rejected, if it defines some field that is not yet supported by the cluster's Prometheus version.

**Review by mviswanathsai — CHANGES_REQUESTED** [review]

**Comment by mviswanathsai** [comment]
I think you would also need to do something similar to the change in #7239

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 4 above)
Comment (heliapb) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:713]: Added thanks

**Review by mviswanathsai — CHANGES_REQUESTED** [review]
Hey @heliapb !
Thank you for addressing my previous feedback in your PR. However, I would like to reiterate my main concern regarding the structure of the e2e test cases. Specifically, I believe we can make the test cases more granular and verbose to improve clarity and maintainability. 

Instead of grouping all the empty or invalid fields into a single test case, I recommend breaking them down into individual test cases for each field. This ensures each test case is specific and focuses on validating one scenario at a time. This approach makes debugging easier and ensures edge cases are covered comprehensively.

For instance, for a field like `SubscriptionID`, we should have:
1. A test where it is correctly defined.
2. A test where it is missing.
3. A test where it is invalid (e.g., an empty string, invalid UUID).

Below is an example illustrating the difference:

### Current Test Case
```go
{
	name: "Invalid AzureSD Missing SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				Environment:          ptr.To("AzurePublicCloud"),
				AuthenticationMethod: ptr.To(monitoringv1alpha1.AuthMethodTypeSDK),
				TenantID:             ptr.To("22222222-2222-2222-2222-222222222222"),
			},
		},
	},
	expectedError: true,
},
```

### Suggested Granular Tests
```go
// Test where SubscriptionID is correctly defined
{
	name: "Valid AzureSD With SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				SubscriptionID: "11111111-1111-1111-1111-111111111111",
			},
		},
	},
	expectedError: false,
},

// Test where SubscriptionID is missing
{
	name: "Invalid AzureSD Missing SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				Environment:          ptr.To("AzurePublicCloud"),
				AuthenticationMethod: ptr.To(monitoringv1alpha1.AuthMethodTypeSDK),
			},
		},
	},
	expectedError: true,
},

// Test where SubscriptionID is invalid. In this case, empty.
{
	name: "Invalid AzureSD With Invalid SubscriptionID",
	scrapeConfigSpec: monitoringv1alpha1.ScrapeConfigSpec{
		AzureSDConfigs: []monitoringv1alpha1.AzureSDConfig{
			{
				SubscriptionID: "",
			},
		},
	},
	expectedError: true,
},
```

Its a bit more work, but as I said, it stays in line with how we have defined our other testcases. There are a few unresolved comments that seem to have missed your attention, please take a look at them as well.

**Comment by heliapb** [comment]
Hi @mviswanathsai sorry I've had missed your last comments, fixed those. Also added the more granular tests, let me know if anything else needs fix, and thanks for the feedback, much appreciated.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 5]
```diff
@@ -656,43 +656,86 @@ type EC2SDConfig struct {
...
+	// +kubebuilder:validation:MinLength=1
 	// +optional
 	ClientID *string `json:"clientID,omitempty"`
 	// Optional client secret. Only required with the OAuth authentication method.
 	// +optional
 	ClientSecret *corev1.SecretKeySelector `json:"clientSecret,omitempty"`
 	// Optional resource group name. Limits discovery to this resource group.
+	// Requires  Prometheus v2.35.0 and above
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MinLength=1
```
[the remark this answers is not in the document]
Comment (heliapb) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Sorry I had missed those reviews, fixed

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 6]
```diff
@@ -656,43 +656,86 @@ type EC2SDConfig struct {
...
 	// +required
 	SubscriptionID string `json:"subscriptionID"`
 	// Optional tenant ID. Only required with the OAuth authentication method.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MinLength=1
 	// +optional
 	TenantID *string `json:"tenantID,omitempty"`
 	// Optional client ID. Only required with the OAuth authentication method.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MinLength=1
```
[the remark this answers is not in the document]
Comment (heliapb) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go]: Sorry I had missed those reviews, fixed

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 7]
```diff
@@ -656,43 +656,86 @@ type EC2SDConfig struct {
...
-	// +kubebuilder:validation:Enum=OAuth;ManagedIdentity;SDK
 	// +optional
-	AuthenticationMethod *string `json:"authenticationMethod,omitempty"`
+	AuthenticationMethod *AuthenticationMethodType `json:"authenticationMethod,omitempty"`
 	// The subscription ID. Always required.
 	// +kubebuilder:validation:MinLength=1
 	// +required
 	SubscriptionID string `json:"subscriptionID"`
 	// Optional tenant ID. Only required with the OAuth authentication method.
+	// +kubebuilder:validation:MinLength=1
```
[the remark this answers is not in the document]
Comment (heliapb) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:688]: Sorry it was accidental, fixed thanks

**Review by mviswanathsai — APPROVED** [review]
LGTM.
CC: @slashpai

**Comment by slashpai** [comment]
@heliapb could you please rebase your PR, after GCE PR merge there is conflict

**Comment by heliapb** [comment]
Hi @slashpai rebase done, thanks

**Review by slashpai — APPROVED** [review]
Great work, Thank you @heliapb and @mviswanathsai  :)
